src/Shared/TagTransformerJsonHelper.cs

```diff
@@ -0,0 +1,192 @@
+// <copyright file="TagTransformerJsonHelper.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System.Buffers;
+using System.Text;
+using System.Text.Json;
+
+namespace OpenTelemetry.Internal;
+
+#nullable enable
+
+internal static class TagTransformerJsonHelper
+{
+    internal static string JsonSerializeArrayTag(Array array)
+        => array switch
+        {
+            // This switch needs to support the same set of types as those in TagTransformer.TransformArrayTagInternal
```

Comment:
I think it should be possible to avoid all this switching if you just use `JsonSerializer.Serialize` with the array value upcast to `object` (since `object` types are aways serialized polymorphically). Assuming you wanted to avoid `JsonSerializer` in order to customize what gets written for each array type, you could replace your `writeAction` delegate with a custom converter that you register in your serializer configuration.

Reply:
See for the discussion which led to this: https://github.com/open-telemetry/opentelemetry-dotnet/issues/4648
Ideally this would use source generators, but the goal is to avoid upgrading the `System.Text.Json` package reference to 6.0. Thus we can't use source generator. That means that I can't use `JsonSerializer` without it being AOT incompatible, so I coded this which avoids it completely.

Reply:
Assuming you only need to serialize arrays of specific types, registering custom converters for the relevant types should ensure that everything works in AOT even with the reflecton-based APIs. For example, the console app only uses pre-net6.0 STJ APIs but works just fine in AOT:

```C#
var options = new JsonSerializerOptions
{
    Converters = 
    { 
        new MyArrayConverter<int>(static (writer, i) => writer.WriteNumberValue(i))
    }
};

Console.WriteLine(JsonSerializer.Serialize(new int[] { 1, 2, 3 }, options));

public class MyArrayConverter<T> : JsonConverter<T[]>
{
    private readonly Action<Utf8JsonWriter, T> _writeAction;

    public MyArrayConverter(Action<Utf8JsonWriter, T> writeAction)
        => _writeAction = writeAction;

    public override T[]? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
        => throw new NotImplementedException();

    public override void Write(Utf8JsonWriter writer, T[] value, JsonSerializerOptions options)
    {
        writer.WriteStartArray();
        foreach (T item in value)
        {
            _writeAction(writer, item);
        }
        writer.WriteEndArray();
    }
}
```

Reply:
The `JsonSerializer.Serialize` method will still warn but with the above configuration it's a false positive that can be suppressed.

Reply:
That would fit option 3 from https://github.com/open-telemetry/opentelemetry-dotnet/issues/4648. I agree it would work, but it's fragile and effectively relies on internal workings of `System.Text.Json`. The main advantage would be to reuse the buffer pooling logic in System.Text.Json, which would be nice, but I'm not sure it's worth the cost of possible future risks.